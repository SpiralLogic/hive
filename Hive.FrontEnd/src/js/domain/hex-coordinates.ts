export type HexCoordinates = {
    readonly q: number;
    readonly r: number;
};
