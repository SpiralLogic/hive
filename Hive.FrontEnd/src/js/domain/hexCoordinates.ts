export interface HexCoordinates {
  q: number;
  r: number;
}