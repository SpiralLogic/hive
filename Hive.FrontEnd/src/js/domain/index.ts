export * from './cell';
export * from './game-state';
export * from './move';
export * from './tile';
export * from './player';
export * from './cell';
export * from './hex-coordinates';
