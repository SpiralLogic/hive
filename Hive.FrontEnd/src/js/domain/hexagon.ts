import {Tile} from './tile';
import {HexCoordinates} from './hex-coordinates';

export interface Hexagon {
    coordinates: HexCoordinates;
    tiles: Tile[];
}
