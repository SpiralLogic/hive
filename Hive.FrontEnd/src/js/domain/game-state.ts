import { Hexagon } from './hexagon';
import { Player } from './player';

export interface GameState {
    hexagons: Hexagon[];
    players: Player[];
}
