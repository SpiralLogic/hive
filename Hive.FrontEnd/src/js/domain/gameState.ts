import { Hexagon } from './hexagon';
import { Player, PlayerId } from './player';

export interface GameState {
  hexagons: Hexagon[];
  players: Player[];
}