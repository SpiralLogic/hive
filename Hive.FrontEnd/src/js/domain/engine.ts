import { GameState } from './game-state';
import { Move } from './move';

export interface HexEngine {
    newGame(): Promise<GameState>;
    move(move: Move): Promise<GameState>;
}
