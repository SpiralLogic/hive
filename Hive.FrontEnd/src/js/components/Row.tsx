import * as React from 'react';
import {Hexagon, HexCoordinates} from '../domain';
import CellMemo from './Cell';

const cellKey = ({q, r}: HexCoordinates) => `${q}-${r}`;
const cellComponent = (cell: Hexagon) => <CellMemo key={cellKey(cell.coordinates)} {...cell} />;
const emptyCell = (key: number) => <div key={key} className="hidden"/>;

type Props = {
    id: number;
    row: Array<Hexagon | false>;
};

function Row(props: Props) {
    const {row} = props;
    const cells = row.map((cell, i) => (cell && cellComponent(cell)) || emptyCell(i));
    return <div className="hex-row">{cells}</div>;
}

Row.displayName = 'Row';

function haveCellTilesBeenUpdated(prevProps: Props, newProps: Props) {
    return prevProps.row.every((cell, i) => newProps.row[i] && cell && (newProps.row[i] as Hexagon).tiles.length == (cell as Hexagon).tiles.length);
}

const RowMemo = React.memo(Row, (p, n) =>
    p.row.length == n.row.length &&
    haveCellTilesBeenUpdated(p, n));

export default RowMemo;
