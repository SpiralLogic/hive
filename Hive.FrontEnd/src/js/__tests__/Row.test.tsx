import * as React from 'react';
import {render, unmountComponentAtNode} from 'react-dom';
import Row from '../components/Row';
import {Hexagon, HexCoordinates, Tile} from '../domain';
import {create} from 'react-test-renderer';

const cells: (Hexagon | false)[] = [{coordinates: {q: 0, r: 1}, tiles: [] as Tile[]}, {coordinates: {q: 1, r: 1}, tiles: [] as Tile[]}, false];
const rowJsx = <Row id={1} row={cells}/>;
let container: HTMLDivElement;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    render(rowJsx, container);
});

test('row has class', () => {
    const row = container.firstElementChild;

    expect(row?.classList).toContain('hex-row');
});

test('row renders multiple cells', () => {
    const cells = document.querySelectorAll('.hex-row > div');

    expect(cells).toHaveLength(3);
});

test('row renders hidden div for missing cells', () => {
    const hiddenCell = document.querySelectorAll('.hidden');

    expect(hiddenCell).toHaveLength(1);
});

test('tile is memoized', () => {
    const first = React.createElement(Row, {id: 1, row: cells});
    const second = React.createElement(Row, {id: 1, row: cells});

    expect(first).toStrictEqual(second);
});

test('tile is re-rendered when cells have changed', () => {
    const first = React.createElement(Row, {id: 1, row: cells});
    const second = React.createElement(Row, {id: 1, row: cells.slice(1)});

    expect(first).not.toStrictEqual(second);
});

test('tile is re-rendered when cell tiles have changed', () => {
    const cell = {coordinates: {q: 0, r: 1}, tiles: [] as Tile[]};
    const first = React.createElement(Row, {id: 1, row: [cell]});

    const cell2 = {coordinates: {q: 0, r: 1}, tiles: [{id:1, playerId:1, content:'',availableMoves:[] as HexCoordinates[]}] };
    const second = React.createElement(Row, {id: 1, row: [cell2]});

    expect(first).not.toStrictEqual(second);
});

test('snapshot', () => {
    expect(create(rowJsx).toJSON()).toMatchSnapshot();
});

afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
});