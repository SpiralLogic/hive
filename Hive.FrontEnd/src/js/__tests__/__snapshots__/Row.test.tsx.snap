// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="hex-row"
>
  <div
    className="hex cell"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
  />
  <div
    className="hex cell"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
  />
  <div
    className="hidden"
  />
</div>
`;
